drivers/flash/flash_stm32_qspi.c

```diff
@@ -0,0 +1,927 @@
+}
+
+/*
+ * Timeout callback.
+ */
+void HAL_QSPI_TimeOutCallback(QSPI_HandleTypeDef *hqspi)
+{
+	struct flash_stm32_qspi_data *dev_data =
+		CONTAINER_OF(hqspi, struct flash_stm32_qspi_data, hqspi);
+
+	LOG_DBG("Enter");
+
+	dev_data->cmd_status = -EIO;
+
+	k_sem_give(&dev_data->sync);
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void flash_stm32_qspi_pages_layout(const struct device *dev,
+				const struct flash_pages_layout **layout,
+				size_t *layout_size)
+{
+	struct flash_stm32_qspi_data *dev_data = DEV_DATA(dev);
+
+	*layout = &dev_data->layout;
+	*layout_size = 1;
+}
+#endif
+
+static const struct flash_driver_api flash_stm32_qspi_driver_api = {
+	.read = flash_stm32_qspi_read,
+	.write = flash_stm32_qspi_write,
+	.erase = flash_stm32_qspi_erase,
+	.write_protection = flash_stm32_qspi_write_protection_set,
+	.get_parameters = flash_stm32_qspi_get_parameters,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = flash_stm32_qspi_pages_layout,
+#endif
+};
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static int setup_pages_layout(const struct device *dev)
+{
+	const struct flash_stm32_qspi_config *dev_cfg = DEV_CFG(dev);
+	struct flash_stm32_qspi_data *data = DEV_DATA(dev);
+	const size_t flash_size = dev_cfg->flash_size;
+	const uint32_t layout_page_size = 65536;
+	uint8_t exp = 0;
+	int rv = 0;
+
+	/* Find the smallest erase size. */
+	for (size_t i = 0; i < ARRAY_SIZE(data->erase_types); ++i) {
+		const struct jesd216_erase_type *etp = &data->erase_types[i];
+
+		if ((etp->cmd != 0)
+		    && ((exp == 0) || (etp->exp < exp))) {
+			exp = etp->exp;
+		}
+	}
+
+	if (exp == 0) {
+		return -ENOTSUP;
+	}
+
+	uint32_t erase_size = BIT(exp);
+
+	/* Error if layout page size is not a multiple of smallest
+	 * erase size.
+	 */
+	if ((layout_page_size % erase_size) != 0) {
+		LOG_ERR("layout page %u not compatible with erase size %u",
+			layout_page_size, erase_size);
+		return -EINVAL;
+	}
+
+	/* Warn but accept layout page sizes that leave inaccessible
+	 * space.
+	 */
+	if ((flash_size % layout_page_size) != 0) {
+		LOG_INF("layout page %u wastes space with device size %zu",
+			layout_page_size, flash_size);
+	}
+
+	data->layout.pages_size = layout_page_size;
+	data->layout.pages_count = flash_size / layout_page_size;
+	LOG_DBG("layout %u x %u By pages", data->layout.pages_count,
+					   data->layout.pages_size);
+
+	return rv;
+}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+static int spi_nor_process_bfp(const struct device *dev,
+			       const struct jesd216_param_header *php,
+			       const struct jesd216_bfp *bfp)
+{
+	const struct flash_stm32_qspi_config *dev_cfg = DEV_CFG(dev);
+	struct flash_stm32_qspi_data *data = DEV_DATA(dev);
+	struct jesd216_erase_type *etp = data->erase_types;
+	const size_t flash_size = jesd216_bfp_density(bfp) / 8U;
+
+	if (flash_size != dev_cfg->flash_size) {
+		LOG_ERR("Unexpected flash size: %u", flash_size);
+	}
+
+	LOG_INF("%s: %u MiBy flash", dev->name, (uint32_t)(flash_size >> 23));
```

Comment:
Should it not be flash->size >> 20 ?

Reply:
No.  Flash sizes for SPI NOR devices are specified in bits.  Long, boring story.

Reply:
Well, I agree this does not match with "%u MiBy flash"

Reply:
Misled again by context: the division by 8 is a couple lines up in the original SPI NOR driver, so yes, it should be >> 20.

Reply:
#31105, thanks for catching this @fzawadiak